### PR TITLE
refactor: cleanup the block production flow logic

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Make the code easy to maintain and debug. 

**Description**

Make the implementation more testable. 

- Removed the usage of callbacks with promises 
- Added separate flows for the each case
- Added the clear logs for each flow
- Splited validator APIs endpoints to independent functions to refactor more easily. 

Related #6159

**Steps to test or reproduce**

- Run all tests

### CODE REVIEW

**Comment by nazarhussain** [comment]
I would recommend the reviewers to review it locally not via Github.

**Comment by nflaig** [comment]
What's reason creating a separate file for each method, don't we lose type safety here, what if a new method is added in routes, will it detect if the implementation is missing on the server?

**Review by g11tech — CHANGES_REQUESTED** [review]
quite an extensive refactor involving critical elements (race promises with cutoff)
blocking for now till can give indepth review

**Review by g11tech — CHANGES_REQUESTED** [review]
because everything is now moved it becomes difficult to see what is the actual diff in logic, can you mark functions with your comment in the PR which haven't changed and have been just relocated

infact would suggest separate PRs for logic change and for relocation to make it easy and also for us to decide what we want or don't want (and build those PRs on top of each other)

**Comment by nazarhussain** [comment]
> because everything is now moved it becomes difficult to see what is the actual diff in logic, can you mark functions with your comment in the PR which haven't changed and have been just relocated
> 
> infact would suggest separate PRs for logic change and for relocation to make it easy and also for us to decide what we want or don't want (and build those PRs on top of each other)

I thought it was clear so didn't mentioned in the PR (sorry for that). 

- Every endpoint is just moved to it's own file without any code change at all. So skip those completely. 
- You must review only one file [src/api/impl/validator/endpoints/produceBlockV3.ts](https://github.com/ChainSafe/lodestar/pull/6212/files#diff-a1e02d1d09e41d544bd0519e8c2f73082e85694d879caed42a59bcbaa2906061) It contains the refactoring of engine/builder flow. 
- The changes in the [packages/utils/src/promise.ts](https://github.com/ChainSafe/lodestar/pull/6212/files#diff-3777f7173ed9d43306b883992b9b3d2062a548ff9b933ca80012ef1516d0eeea) is very straight forward to simplify the flow and can be verified through the test cases. 

Hope it will be clear now the scope of the review.

File: packages/utils/test/unit/promiserace.test.ts [hunk 1]
```diff
@@ -14,91 +15,67 @@ describe("racePromisesWithCutoff", () => {
...
-  // Second item in testcase row i.e. array of numbers represent delay at which promise to be resolved
-  // or rejected (-ve number)
-  // Third item in testcase row is the expected value or error message in string
-  // Last item is expected events
-  const testcases: [string, number[], (string | Error)[], RaceEvent[]][] = [
-    ["all resolve pre-cutoff", [100, 200], ["100", "200"], [resolved, resolved, precutoff]],
-    ["all resolve/reject pre-cutoff", [100, -200], ["100", "-200"], [resolved, rejected, precutoff]],
-    ["all reject pre-cutoff", [-100, -200], ["-100", "-200"], [rejected, rejected, precutoff]],
-    ["all reject pre-timeout", [-1100, -1200], ["-1100", "-1200"], [cutoff, rejected, rejected, pretimeout]],
-    ["race and resolve pre-timeout", [1100, 1200], ["1100", "pending"], [cutoff, resolved, pretimeout]],
```
Comment (g11tech) [inline packages/utils/test/unit/promiserace.test.ts:35]: why have the raceevents been removed here?

Comment (nazarhussain) [reply packages/utils/test/unit/promiserace.test.ts:35]: Mixing event callbacks with the Promise is not a good pattern, as you will review the new implementation is much clean and achieve what is needed without the events.

**Comment by g11tech** [comment]
can you again divide this PR into further parts:

1. where you only optimize for the body
2. your proposed race promises refactor

please build them on top of either `blinded-local` branch PR (https://github.com/ChainSafe/lodestar/pull/6227) and would suggest not renaming functions yet.

**Comment by nazarhussain** [comment]
Closing in favour of https://github.com/ChainSafe/lodestar/pull/6241

**Pull request closed without merging**